### PR TITLE
express-note-api | Paul Park

### PR DESCRIPTION
**Pull request by paulpark330**
express-notes-api

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
+const express = require('express');
+const app = express();
+const notebookData = require('./data.json');
+const fs = require('fs');
+
+const notebook = notebookData.notes;
+const noteId = notebookData.nextId;
```
Comment (thebearingedge) [inline express-notes-api/index.js:7]: This is causing your server to erase the previously saved note every time. Try just using `notebook.nextId` directly in your POST route.

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    notebookData.notes[noteId] = newNote;
+    notebookData.nextId++;
+    const notebookJSON = JSON.stringify(notebookData, null, 2);
+    fs.writeFile('data.json', notebookJSON, 'utf8', err => {
+      if (err) {
+        const error = { error: 'An unexpected error occurred' };
+        res.status(500).send(error);
+      }
+    });
+    res.status(201).send(newNote);
```
Comment (thebearingedge) [inline express-notes-api/index.js:50]: Be careful not to respond until the writeFile completes. Check your other uses of writeFile too.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js (hunk 2 above)
Comment (paulpark330) [reply express-notes-api/index.js:50]: Hey Tim, I just want clarification on this comment!
`res.status(201).send(newNote);`
If this line is outside of the writeFile code block, shouldn't the response be after the writeFile function completes?

Comment (thebearingedge) [reply express-notes-api/index.js:49]: @paulpark330 The only way to make sure that something happens after an asynchronous operation is to put it in the callback's code block.

Comment (paulpark330) [reply express-notes-api/index.js:49]: Definitely forgot about the asynchronous part of the writeFile method.
I will make more adjustments!
Thank you

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    };
+    notebookData.notes[notebookData.nextId] = newNote;
+    notebookData.nextId++;
+    const notebookJSON = JSON.stringify(notebookData, null, 2);
+    fs.writeFile('data.json', notebookJSON, 'utf8', err => {
+      if (err) {
+        const error = { error: 'An unexpected error occurred' };
+        res.status(500).send(error);
+      }
+      res.status(201).send(newNote);
```
Comment (thebearingedge) [inline express-notes-api/index.js:48]: Be sure to use `else` or `return` to prevent the server from attempting to respond twice to the same request.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

**Review by thebearingedge — APPROVED** [review]

**Pull request merged**